File: 3733-length-of-longest-v-shaped-diagonal-segment/length-of-longest-v-shaped-diagonal-segment.java
class Solution {
    public int lenOfVDiagonal(int[][] g) {
        int n = g.length, m = g[0].length;
        int[][] D = { { -1, 1 }, { 1, 1 }, { 1, -1 }, { -1, -1 } };

        int[][][] dp0 = new int[4][n][m];
        int[][][] dp2 = new int[4][n][m];
        int res = 0;

        // Fill DP arrays
        for (int d = 0; d < 4; d++) {
            int dr = D[d][0], dc = D[d][1];
            int[] rows, cols;

            if (dr >= 0) {
                rows = new int[n];
                for (int i = 0; i < n; i++)
                    rows[i] = n - 1 - i;
            } else {
                rows = new int[n];
                for (int i = 0; i < n; i++)
                    rows[i] = i;
            }

            if (dc >= 0) {
                cols = new int[m];
                for (int j = 0; j < m; j++)
                    cols[j] = m - 1 - j;
            } else {
                cols = new int[m];
                for (int j = 0; j < m; j++)
                    cols[j] = j;
            }

            for (int i : rows)
                for (int j : cols) {
                    int ni = i + dr, nj = j + dc;
                    if (g[i][j] == 0)
                        dp0[d][i][j] = 1 + (inb(ni, nj, n, m) ? dp2[d][ni][nj] : 0);

                    if (g[i][j] == 2)
                        dp2[d][i][j] = 1 + (inb(ni, nj, n, m) ? dp0[d][ni][nj] : 0);
                }
        }

        // Try every start with "1"
        for (int d = 0; d < 4; d++) {
            int dr = D[d][0], dc = D[d][1];
            int t = (d + 1) % 4;

            for (int i = 0; i < n; i++)
                for (int j = 0; j < m; j++) {
                    if (g[i][j] != 1)
                        continue;
                    int r = i, c = j, need = 1, L = 0;

                    while (inb(r, c, n, m) && g[r][c] == need) {
                        L++;
                        res = Math.max(res, L);

                        // Try turning
                        int tr = r + D[t][0], tc = c + D[t][1];
                        if (inb(tr, tc, n, m))
                            if ((L & 1) == 1)
                                res = Math.max(res, L + dp2[t][tr][tc]);
                            else
                                res = Math.max(res, L + dp0[t][tr][tc]);

                        r += dr;
                        c += dc;
                        if (need == 1)
                            need = 2;
                        else if (need == 2)
                            need = 0;
                        else
                            need = 2;
                    }
                }
        }

        return res;
    }

    private boolean inb(int i, int j, int n, int m) {
        return i >= 0 && i < n && j >= 0 && j < m;
    }
}

File: 3733-length-of-longest-v-shaped-diagonal-segment/length-of-longest-v-shaped-diagonal-segment.go
func lenOfVDiagonal(g [][]int) int {
	n, m := len(g), len(g[0])
	D := [][2]int{{-1, 1}, {1, 1}, {1, -1}, {-1, -1}} // 4 diagonal directions
	inb := func(i, j int) bool { return i >= 0 && i < n && j >= 0 && j < m }

	res := 0

	// dp0[d][i][j] = alternating length starting with "0"
	// dp2[d][i][j] = alternating length starting with "2"
	dp0 := make([][][]int, 4)
	dp2 := make([][][]int, 4)
	for d := 0; d < 4; d++ {
		dp0[d] = make([][]int, n)
		dp2[d] = make([][]int, n)
		for i := range dp0[d] {
			dp0[d][i] = make([]int, m)
			dp2[d][i] = make([]int, m)
		}
	}

	// Fill DP arrays
	for d, dir := range D {
		dr, dc := dir[0], dir[1]
		rows := make([]int, 0)
		cols := make([]int, 0)

		if dr >= 0 {
			for i := n - 1; i >= 0; i-- {
				rows = append(rows, i)
			}
		} else {
			for i := 0; i < n; i++ {
				rows = append(rows, i)
			}
		}

		if dc >= 0 {
			for j := m - 1; j >= 0; j-- {
				cols = append(cols, j)
			}
		} else {
			for j := 0; j < m; j++ {
				cols = append(cols, j)
			}
		}

		for _, i := range rows {
			for _, j := range cols {
				ni, nj := i+dr, j+dc
				if g[i][j] == 0 {
					if inb(ni, nj) {
						dp0[d][i][j] = 1 + dp2[d][ni][nj]
					} else {
						dp0[d][i][j] = 1
					}
				}
				if g[i][j] == 2 {
					if inb(ni, nj) {
						dp2[d][i][j] = 1 + dp0[d][ni][nj]
					} else {
						dp2[d][i][j] = 1
					}
				}
			}
		}
	}

	// Try every starting point with "1"
	for d, dir := range D {
		dr, dc := dir[0], dir[1]
		t := (d + 1) % 4 // clockwise turn

		for i := 0; i < n; i++ {
			for j := 0; j < m; j++ {
				if g[i][j] != 1 {
					continue
				}
				r, c, need, L := i, j, 1, 0
				for inb(r, c) && g[r][c] == need {
					L++
					if L > res {
						res = L
					}

					// Try turning clockwise
					tr, tc := r+D[t][0], c+D[t][1]
					if inb(tr, tc) {
						if L%2 == 1 {
							res = max(res, L+dp2[t][tr][tc])
						} else {
							res = max(res, L+dp0[t][tr][tc])
						}
					}

					// Move forward
					r, c = r+dr, c+dc
					if need == 1 {
						need = 2
					} else if need == 2 {
						need = 0
					} else {
						need = 2
					}
				}
			}
		}
	}

	return res
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}